// @heroui/react/dist/components/fieldset/fieldset.js
"use client";
import { fieldsetVariants } from '@heroui/styles';
import React__default, { createContext, useContext } from 'react';
import { Provider, RadioGroupContext, SliderContext } from 'react-aria-components';
import { composeSlotClassName } from '../../utils/compose.js';
import { dom } from '../../utils/dom.js';
import { jsx } from 'react/jsx-runtime';

const FieldsetContext = /*#__PURE__*/createContext({});

/* -------------------------------------------------------------------------------------------------
 * Fieldset Root
 * -----------------------------------------------------------------------------------------------*/

const FieldsetRoot = ({
  children,
  className,
  ...props
}) => {
  const slots = React__default.useMemo(() => fieldsetVariants({}), []);

  // Mirror native `<fieldset disabled>` as `data-disabled="true"` so the
  // existing `[data-disabled="true"] .label` (and similar) ancestor selectors
  // cascade disabled styling to descendant fields, just like a direct
  // `isDisabled` prop on TextField/Checkbox/etc. would.
  const isDisabled = "disabled" in props && props.disabled === true;
  return /*#__PURE__*/jsx(FieldsetContext, {
    value: {
      slots
    },
    children: /*#__PURE__*/jsx(dom.fieldset, {
      className: slots?.base({
        className
      }),
      "data-disabled": isDisabled || undefined,
      "data-slot": "fieldset",
      ...props,
      children: isDisabled ?
      /*#__PURE__*/
      // Slider and RadioGroup render as <div> (no native form controls),
      // so the browser doesn't propagate <fieldset disabled> to them.
      // Forward `isDisabled` explicitly via React Aria's context so their
      // wrapper elements receive `data-disabled` and behave as disabled.
      jsx(Provider, {
        values: [[RadioGroupContext, {
          isDisabled: true
        }], [SliderContext, {
          isDisabled: true
        }]],
        children: children
      }) : children
    })
  });
};

/* -------------------------------------------------------------------------------------------------
 * Fieldset Legend
 * -----------------------------------------------------------------------------------------------*/

const FieldsetLegend = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(FieldsetContext);
  return /*#__PURE__*/jsx(dom.legend, {
    className: composeSlotClassName(slots?.legend, className),
    "data-slot": "fieldset-legend",
    ...props
  });
};

/* -------------------------------------------------------------------------------------------------
 * Field Group
 * -----------------------------------------------------------------------------------------------*/

const FieldGroup = ({
  className,
  ...rest
}) => {
  const {
    slots
  } = useContext(FieldsetContext);
  return /*#__PURE__*/jsx(dom.div, {
    className: composeSlotClassName(slots?.fieldGroup, className),
    "data-slot": "fieldset-field-group",
    ...rest
  });
};

/* -------------------------------------------------------------------------------------------------
 * Field Actions
 * -----------------------------------------------------------------------------------------------*/

const FieldsetActions = ({
  children,
  className,
  ...rest
}) => {
  const {
    slots
  } = useContext(FieldsetContext);
  return /*#__PURE__*/jsx(dom.div, {
    className: composeSlotClassName(slots?.actions, className),
    "data-slot": "fieldset-actions",
    ...rest,
    children: children
  });
};

export { FieldGroup, FieldsetActions, FieldsetLegend, FieldsetRoot };
